### PR TITLE
Improve Ephemeral Onion Service private key handling

### LINKED ISSUE #190: When the ephemeral hs is initialized with an existing key ehs.private and ehs.hostname remains uninitialized
Right now when the when the ephemeral hs is initialized with an existing key ehs.private and ehs.hostname remains uninitialized.

In order to reproduce and unit tests the fix the following code could be used:

```python
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives.asymmetric import rsa
from cryptography.hazmat.primitives import serialization

private_key = rsa.generate_private_key(
    public_exponent=65537,
    key_size=1024,
    backend=default_backend()
)

privkey = private_key.private_bytes(
   encoding=serialization.Encoding.PEM,
   format=serialization.PrivateFormat.TraditionalOpenSSL,
   encryption_algorithm=serialization.NoEncryption()
)

privkey = privkey.replace('-----BEGIN RSA PRIVATE KEY-----', '')
privkey = privkey.replace('-----END RSA PRIVATE KEY-----', '')
privkey = privkey.replace('\n', '')
privkey = 'RSA1024:' + privkey

hs = txtorcon.EphemeralHiddenService(["80 127.0.0.1:8082"], privkey)
yield hs.add_to_tor(proto.tor_protocol)
```

it will be then possible to verify that when the callbacks are resolved hs.private_key and hs.hostname will be still undefined resulting in a KeyError if one access them.

### PR DESCRIPTION
**Pull request by felipedau**
These changes seem to fix #190 and implement a different check for the type and blob provided by the user.

Right now it only checks if the key contains a `':'` and in case the type is not `'NEW'`, makes sure the blob is not empty. Should we make sure both the type and blob are not empty? Also, as an extra `AttributeError` is being catched in case the user does not provide a string (well, something that does not `split`).

I am not sure about this, but I think it is a good practice to declare all attributes in `__init__`. Is it okay to declare `hostname` and `private_key` there (by just setting them to `None`)?

We should probably add tests for these too.

Let me know what you think. Thanks!

### CODE REVIEW

File: txtorcon/torconfig.py [hunk 1]
```diff
@@ -445,9 +445,13 @@ def __init__(self, ports, key_blob_or_type='NEW:BEST', auth=[], ver=2):
         # FIXME nicer than assert, plz
         assert ' ' not in self._key_blob
         assert isinstance(ports, list)
-        if not key_blob_or_type.startswith('NEW:') \
-           and (len(key_blob_or_type) > 825 or len(key_blob_or_type) < 820):
-            raise RuntimeError('Wrong size key-blob')
+        try:
+            key_type, key_blob = key_blob_or_type.split(':', 1)
+        except (AttributeError, ValueError):
+            raise RuntimeError('Wrong key-blob format')
```
Comment (meejah) [inline txtorcon/torconfig.py]: These should be `ValueErrors` too, IMO.

Personally, I: wouldn't catch `AttributeError` (or, if there is some sort of check basically against the type, check `isinstance(key_blob_or_type, six.text_type)` or whatever is appropriate). I'd also be more explicit about the colon, something like `if not ':' in ...: raise ValueError("key_blob_or_type must be a string like ...")`

Might also be worth putting this all in a helper like `_is_valid_keyblob()`.

File: txtorcon/torconfig.py [hunk 2]
```diff
@@ -461,8 +465,10 @@ def add_to_tor(self, protocol):
         ans = yield protocol.queue_command(cmd)
         ans = find_keywords(ans.split('\n'))
         self.hostname = ans['ServiceID'] + '.onion'
-        if self._key_blob == 'NEW:BEST':
+        if self._key_blob.startswith('NEW:'):
             self.private_key = ans['PrivateKey']
+        else:
+            self.private_key = self._key_blob
```
Comment (meejah) [inline txtorcon/torconfig.py]: :+1:

**Comment by meejah** [comment]
> Should we make sure both the type and blob are not empty?

Yes, probably! I'm not sure if passing no type to Tor is valid or not, but saying it isn't at this level is fine with me.

File: txtorcon/torconfig.py (hunk 1 above)
Comment (felipedau) [inline txtorcon/torconfig.py]: > These should be ValueErrors too, IMO.

Okay. So instead of raising RuntimeErrors, raise ValueErrors?

> Personally, I: wouldn't catch AttributeError (or, if there is some sort of check basically against the type, check isinstance(key_blob_or_type, six.text_type) or whatever is appropriate).

That's certainly a better solution!

> I'd also be more explicit about the colon, something like if not ':' in ...: raise ValueError("key_blob_or_type must be a string like ...")

Hmm, I thought it would be "more Pythonic" to assume that if `split` does not return two values, then the assignment will fail with a ValueError. I will make sure to use a more explicit check then.

**Comment by felipedau** [comment]
> I definitely wouldn't put hostname in, because it is (or should be!) a read-only attribute and the user has no business messing with it.
> 
> There's some argument for putting private_key in, but I would leave it out because I happen to have knowledge from the future ;). With the new Onion APIs, the default services will be ephemeral, and triggered off passing a private_key arg -- which if None will mean "let Tor create a new one" and otherwise is the keyblob. In any case, this is a sort of "opaque" thing -- the user shouldn't really ever create their own, merely echo back one they got by previously asking Tor to create them one.

Hmm, I did not suggest to receive `hostname` and `private_key` in `__init__`, but just declare them there so people know they exist. I remember the first time I was learning how to use the ephemeral stuff and I was confused if I could access those - later I noticed they were declared in the method below.

Currently, users would be able to mess with it after calling `add_to_tor` because they are regular attributes. Would it be more correct to declare (get) properties?

Another question: How important is it to provide the key type/blob in `__init__`? Couldn't it be provided just to `add_to_tor` instead?

File: txtorcon/torconfig.py (hunk 1 above)
Comment (meejah) [inline txtorcon/torconfig.py]: Yeah, that would kind of be more "if it quacks like a duck" etc. But in this case I think checking for a string explicitly is fine. Also, these days "subclassing builtin types", especially str, isn't really a thing you should do anyway.

**Comment by meejah** [comment]
> Hmm, I did not suggest to receive hostname and private_key in __init__, but just declare them there so people know they exist

Ah, I see what you mean. They way they're done via `__getattr__` is very gross. I've kind of been ignoring that as I'm working on getting "the new onion stuff" moved over from `release-1.x`.

Ultimately, onions will be created via a factory-method call on `Tor` in the very near future, and users shouldn't need to instantiate these classes themselves at all (nor use `add_to_tor`). That said, I will keep this one around and people can use it .. so if you want to fix how the hostname etc attributes are accessed that would be great.

**Comment by meejah** [comment]
p.s. it might be better to save ^ for a follow-on PR so we can get these improvements merged when they're ready and so this PR doesn't grow "too big"/long-running etc ;)

Trying to learn my lesson there :)

**Comment by felipedau** [comment]
I should definitely take a look at the `release-1-x` code, these are awesome additions.

So, what can I do to make your life easier when merging?

Does adding `_is_valid_keyblob` result in conflicts?

It is not very clear to me what to do now and later. Here is what I got so far:

- In this PR: improve the blob/type checks and add tests for them as well as the persistent key and hostname.

- In a new PR (after the new features are merged): make the key and hostname private and provide only read access to them.

Is that right?

**Comment by meejah** [comment]
yes, that sounds right. don't worry about conflicts, adding a helper is fine.

**Comment by felipedau** [comment]
> yes, that sounds right. don't worry about conflicts, adding a helper is fine.

Cool! Thanks :)

I just amended that last check. I decided to use the regex `[^ :]+:[^ :]+$`. What do you think? I will then add the `isinstance` check to that conditional too.

I will move everything to the `_is_valid_keyblob`. Do you think that call should be the first instruction in the `__init__` instead of assigning the attributes?

**Comment by felipedau** [comment]
Update:

Added the `isinstance` check and moved to `_is_valid_keyblob`. Let me know if you think it is a good idea to use a `@classmethod`.

Also fixed the wrong key-blob test expecting for a `RuntimeError` and added a few more wrong key-blobs to the test. Do you use anything (parametrization/fuzzy tests) that could make that test better?

Created another test to make sure the hostname a private key are persisted when the EOS is initialized with an existing key.

Let me know if you think the changes look like a good idea.

P.S. I took a quick look at the `release-1.x` branch and there are already properties for the hostname and key, so I think that second PR we discussed above is unnecessary.

Thanks!

File: txtorcon/torconfig.py [hunk 3]
```diff
@@ -423,6 +424,11 @@ class EphemeralHiddenService(object):
     https://gitweb.torproject.org/torspec.git/tree/control-spec.txt#n1295
     '''
 
+    @classmethod
+    def _is_valid_keyblob(cls, key_blob_or_type):
+        return (isinstance(key_blob_or_type, str) and
+                re.match(r'[^ :]+:[^ :]+$', key_blob_or_type))
```
Comment (meejah) [inline txtorcon/torconfig.py]: Should this be `six.text_type`? or should it be `bytes`?

Currently, the only existing thing happens to be ASCII, leading me to believe that `text_type` is more-correct. Also, it'll typically be "stuff a user types in" also leading to `text_type`...that'll mean it demands `unicode` on Py2.

**Comment by meejah** [comment]
great, thanks!

File: txtorcon/torconfig.py (hunk 3 above)
Comment (felipedau) [inline txtorcon/torconfig.py]: Oh man, you mentioned that and I totally forgot! Sorry :/ I agree, I think it should be `six.text_type`.

Would you like me to fix that? Will it break stuff because people would then have to use `u'ALG:KEY'`?

Comment (meejah) [inline txtorcon/torconfig.py]: Yeah, it does kind of break things, unicode-wise. Maybe what we should do is use the "nativeString" (I think it's called?) helper from Twisted.

The idea here is: if anyone is right now using "`NEW:BEST`" or similar (i.e. non-unicode on Py2) then we can try to upgrade it to unicode by pretending it's ASCII (`blob.decode('ascii')`). If that works, great! Otherwise, no error. Then this works for people using the "meejah didn't think about unicode properly" API already...

So, basically: a helper in `util` that "maybe converts to unicode, or throws an exception" given some input-thing. Or just use `twisted.python.compat.nativeString`...

Comment (felipedau) [inline txtorcon/torconfig.py]: I prefer this approach to not break stuff rather than the other - not at least without a warning :P

I also prefer using what Twisted already offers or maybe creating this abstract helper in `util` that uses `nativeString` underneath, but can be easily modified if one day is necessary.

(That's just my opinion as a user. You know better than me.)

Questions:

- What is this kind of change supposed to do? Always send unicode to Tor?
- Is it something that should be done in the entire code base?
- Is the whole API supposed to receive/return unicode?

Thanks!

Comment (meejah) [inline txtorcon/torconfig.py]: On the wire, we always send bytes to Tor. But "in general" when interacting with users you get and need to deal with unicode. In this particular case, it should always be 7-bit ASCII so we shouldn't run into problems.

Entire-code-base-wise: yes, more looking at the handling of unicode vs bytes is always good. This wasn't very explicit before 0.19.0 release (i.e. Python3 wasn't supported before then). But, fixing just this one would be good as a first step.

Using the Twisted API directly is fine.
